### PR TITLE
router: auto_host_rewrite_header support

### LINKED ISSUE #6486: Upstream request header modification
I would like to have a way of modifying upstream request headers _including_ `host/:authority`.
I can alter headers via `request_headers_to_add` or lua code, but changing `host/:authority` is done before virtual host matching and so it changes virtual host picked, or results in 404.
I need virtual host to be picked using downstream request's `host/:authority` and only then alter headers sent to upstream.

Why do I need it: I would like to multiplex traffic from many services to envoy through single listener with one virtual host per service. Envoy is not aware of all domains (with possible subdomains) of these services, so I would substitute original host with known per service vhost name storing original host in some other header first. Envoy selects exact virtual hosts efficiently with a map and so I would like not to use wildcard domains, or long route chains matching custom headers.

Example of traffic:
```
Original request      | Request from load balancer to Envoy         | Request from Envoy to upstream
------------------------------------------------------------------------------------------------------
Host: a.example.com   | Host: a,  X-Original-Host: a.example.com    | vhost a -> Host: a.example.com
Host: b.example.com   | Host: b,  X-Original-Host: b.example.com    | vhost b -> Host: b.example.com
Host: a.c.example.com | Host: c,  X-Original-Host: a.c.example.com  | vhost c -> Host: a.c.example.com
Host: b.c.example.com | Host: c,  X-Original-Host: b.c.example.com  | vhost c -> Host: b.c.example.com
```
Is it possible?

### PR DESCRIPTION
**Pull request by bartebor**
Description: Make it possible to substitute upstream's request `host`/`:authority` header with value of any (downstream or custom) header. This allows for efficient implementation of more involving deployments with specially configured downstream load balancers. 
Risk Level: Low, does not change default behaviour
Testing: Unit tests added
Docs Changes: Description in proto file
Release Notes: Updated
Fixes #6486

### CODE REVIEW

File: source/common/config/rds_json.cc [hunk 1]
```diff
@@ -275,15 +275,22 @@ void RdsJson::translateRoute(const Json::Object& json_route, envoy::api::v2::rou
 
     JSON_UTIL_SET_STRING(json_route, *action, prefix_rewrite);
 
+    int host_rewrite_options = 0;
```
Comment (zuercher) [inline source/common/config/rds_json.cc]: Please revert this file. We no longer add new features to the v1 API.

File: source/common/json/config_schemas.cc [hunk 2]
```diff
@@ -671,6 +671,7 @@ const std::string Json::Schema::ROUTE_ENTRY_CONFIGURATION_SCHEMA(R"EOF(
       "prefix_rewrite" : {"type" : "string"},
       "host_rewrite" : {"type" : "string"},
       "auto_host_rewrite" : {"type" : "boolean"},
+      "auto_host_rewrite_header" : {"type" : "string"},
```
Comment (zuercher) [inline source/common/json/config_schemas.cc]: Likewise, this schema is for the v1 api and should be reverted.

File: source/common/router/config_impl.cc [hunk 3]
```diff
@@ -513,6 +517,14 @@ void RouteEntryImplBase::finalizeRequestHeaders(Http::HeaderMap& headers,
   vhost_.globalRouteConfig().requestHeaderParser().evaluateHeaders(headers, stream_info);
   if (!host_rewrite_.empty()) {
     headers.Host()->value(host_rewrite_);
+  } else if (auto_host_rewrite_header_) {
+    Http::HeaderEntry* header = headers.get(*auto_host_rewrite_header_);
+    if (header != nullptr) {
+      absl::string_view headerValue = header->value().getStringView();
+      if (!headerValue.empty()) {
+        headers.Host()->value(std::string(headerValue));
```
Comment (zuercher) [inline source/common/router/config_impl.cc]: You can leave out the temporary std::string and just pass `headerValue` directly to
 `HeaderEntry::value(absl::string_view)` which will copy the string.

File: source/common/config/rds_json.cc (hunk 1 above)
Comment (bartebor) [reply source/common/config/rds_json.cc]: File reverted.

File: source/common/json/config_schemas.cc (hunk 2 above)
Comment (bartebor) [reply source/common/json/config_schemas.cc]: File reverted.

File: source/common/router/config_impl.cc (hunk 3 above)
Comment (bartebor) [reply source/common/router/config_impl.cc]: Fixed.

**Comment by bartebor** [comment]
@zuercher, thanks for the review!

File: source/common/router/config_impl.cc [hunk 4]
```diff
@@ -513,6 +517,14 @@ void RouteEntryImplBase::finalizeRequestHeaders(Http::HeaderMap& headers,
   vhost_.globalRouteConfig().requestHeaderParser().evaluateHeaders(headers, stream_info);
   if (!host_rewrite_.empty()) {
     headers.Host()->value(host_rewrite_);
+  } else if (auto_host_rewrite_header_) {
+    Http::HeaderEntry* header = headers.get(*auto_host_rewrite_header_);
+    if (header != nullptr) {
+      absl::string_view headerValue = header->value().getStringView();
```
Comment (mattklein123) [inline source/common/router/config_impl.cc]: nit: `header_value`

Comment (bartebor) [reply source/common/router/config_impl.cc]: My Java background is showing... fixed.

**Pull request merged**